<HEAD>
<TITLE>CHSMiner Help</TITLE>
</HEAD>
<BODY BGCOLOR=white>

<CENTER><H1 style="background:#BBBBBB">CHSMiner</H1></CENTER>
<CENTER><H1 style="width:50%; background:#CCCCCC;">Download</H1></CENTER>
<H3>Downloads</H3>
<UL>	
	<LI>The latest version of CHSMiner program: <A HREF="prog/CHSMiner.jar">CHSMiner 1.1</A></LI>
	<LI>Data sets to test the program: <A HREF="data/paralogs.txt">paralogs</A> and
		<A HREF="data/orthologs.txt">orthologs</A> data set</LI>	
</UL>
<P>
Our <A HREF="#L">License statement</A> contains rights and restrictions 
associated with use of the accompanying software and documentation.
</P>

<H3>News</H3>
<UL>
	<LI>1.1 release - 8/10/10</LI>
	<P>
		Fixed bugs in outputting results for Ensembl data.
	</P>
	<P>	
		Listed matched genes in CHS in output text file.
	</P>
	<LI>Release at Softpedia - 5/25/10</LI> 
	<P>
		CHSMiner was added to Softpedia's database of software programs for Mac OS
		[<A HREF="http://mac.softpedia.com/get/Math-Scientific/CHSMiner.shtml">Softpedia</A>].
	</P>
	<LI>1.0 release - 1/15/09</LI>
	<P>
		The paper about CHSMiner was accepted by <I>Algorithms Mol Biol</I>
		[<A HREF="http://www.ncbi.nlm.nih.gov/pubmed/19146671">PubMed</A>].
	</P>
<UL>

<CENTER><H1 style="width:50%; background:#CCCCCC;">Help</H1></CENTER>
<CENTER><H2><A NAME="INDEX">Index</A></H2></CENTER>
<OL>
	<LI><A HREF="#G">General help for CHSMiner (1.1)</A></LI>
	<LI><A HREF="#F">Data input</A></LI>
	<LI><A HREF="#E">Preprocessing</A></LI>
	<LI><A HREF="#M">CHSs detection and evaluation</A></LI>
	<LI><A HREF="#P">Output files</A></LI>
	<LI><A HREF="#B">CHSs visualization</A></LI>
	<LI><A HREF="#D">More details about statistical evaluation</A></LI>
	<LI><A HREF="#R">References</A></LI>
	<LI><A HREF="#L">License statement</A></LI>
	<LI><A HREF="#C">Contact us</A></LI>
</OL>

<CENTER><H2><A NAME="G">General help for CHSMiner (1.1) </A></H2></CENTER>
<P>
CHSMiner provides an integrated GUI environment for the detection, evaluation 
and visulization of chromosomal homologous segments (CHSs). 
</P>

<P><H3>Biological considerations</H3></P>
<P>
CHS can be identified by genomic regions that contain a set of colinear homologous 
genes (Fig 1A). A limited number of unmatched genes could be allowed in this context because of 
extensive deletion and insersion. However, some studies have shown that conserved gene order
may be too strict for ancient divergence (e.g. whole genome duplication in early vertebrate 
evolution), as gene inversion is another dominant force to dengenerate the genome. In order to 
fully exploit potential homology, CHSMiner define CHS based on shared gene content alone (Fig 1B).
You can see the review of Simillion et al. <A HREF="#1">[1]</A> for more details.
</P>
<CENTER>
	<IMG SRC = "img/CHS.png" WIDTH = 500, HEIGHT = 125, ALT = "Figure 1">
	<P><STRONG>Figure 1 Graphical display of CHS</STRONG></P>
	<P>Homologous genes at two genomic locations are labelled with the same color and linked.</P>
</CENTER>

<P><H3>Features</H3></P>
<UL>
	<LI>Define CHSs based on shared gene content alone to fully explore potential homology.</LI>
	<LI>Uniform design for both syntenies (between two different genomes) and 
	paralogons (within the same genome). 
	</LI>
	<LI>Fast greedy algorithm to search for CHSs and formal statistical evaluation of the results.</LI>
	<LI>Interactive visualization of CHSs from an individual one to whole genome pattern.</LI>
	<LI>Entire Java package and convenient windows interface for end users. </LI>
</UL>
<P><B>Note: </B>
	CHSMiner is particularly useful to identify ancient CHSs that cannot be recognized
	from primary sequences. 
</P>

<P><H3>Getting started</H3></P>
<P>
CHSMiner was entirely written in Java and complied in a jar package.
Computers with Java 1.5 or higher installed are the only necessary condition 
to run the application. You can download the latest Java Runtime Environment 
(JRE) or or Java Development Kit (JDK) from 
<A HREF = "http://java.sun.com/javase/downloads">http://java.sun.com/javase/downloads</A>.
</P>
<P>
You can start the application by either running "java -jar CHSMiner.jar" 
from the command line, or just double-clicking the jar file (Fig 2).
</P>
<P>
<B>Note: </B>It is recommended to run the application by "java -Xms256M -Xmx256M -jar CHSMiner.jar",
which will allocate more memory for it.
</P>
<CENTER>
	<IMG SRC = "img/interface.png", ALT = "Figure 2">
	<P><STRONG>Figure 2 The interface of CHSMiner</STRONG></P>
</CENTER>

<A HREF="#INDEX"> <EM>Back to Index</EM> </A>

<CENTER><H2><A NAME="F">Data input</A></H2></CENTER>
<P>For user convenience, CHSMiner provides two ways for data input:</P>

<H3>Read data from a file</H3>
<P>Click the button <IMG SRC = "icon/file.png", ALT = "file"> to read data from a file.</P>
<P>
Below is an example of the CHSMiner format, including the name and type of each 
column. Every line is a gene record with items separated by Tab. The file can 
also contains titles and other annotations beginning with "#". There is no 
restriction for the order of the records, as the programme will automatically 
set ranks to them in preprocessing.
</P>
<CENTER>
	<TABLE ALIGN=CENTER BORDER=1 CELLSPACING=1 CELLPADDING=5>
		<TR>
			<TD><STRONG># Gene ID</STRONG></TD>
			<TD><STRONG>Family ID</STRONG></TD>
			<TD><STRONG>Symbol</STRONG></TD>
			<TD><STRONG>Species</STRONG></TD>
			<TD><STRONG>Chromosome</STRONG></TD>
			<TD><STRONG>Orient</STRONG></TD>
			<TD><STRONG>Start</STRONG></TD>
			<TD><STRONG>End</STRONG></TD>
		</TR>
		<TR>
			<TD><I>String</I></TD>
			<TD><I>String</I></TD>
			<TD><I>String</I></TD>
			<TD><I>String</I></TD>
			<TD><I>String</I></TD>
			<TD><I>String</I></TD>
			<TD><I>Integer</I></TD>
			<TD><I>Integer</I></TD>
		</TR>
		<TR>
			<TD>104</TD>
			<TD>8280</TD>
			<TD>ADARB1</TD>
			<TD>Homo sapiens</TD>
			<TD>21</TD>
			<TD>+</TD>
			<TD>45318942</TD>
			<TD>45470901</TD>
		</TR>
		<TR>
			<TD>5930</TD>
			<TD>5039</TD>
			<TD>RBBP6</TD>
			<TD>Homo sapiens</TD>
			<TD>16</TD>
			<TD>+</TD>
			<TD>24458408</TD>
			<TD>24491684</TD>
		</TR>
		<TR>
			<TD>110532</TD>
			<TD>8280</TD>
			<TD>Adrab1</TD>
			<TD>Mus musculus</TD>
			<TD>10</TD>
			<TD>-</TD>
			<TD>76753477</TD>
			<TD>76881017</TD>
		</TR>
		<TR>
			<TD>...</TD>
			<TD>...</TD>
			<TD>...</TD>
			<TD>...</TD>
			<TD>...</TD>
			<TD>...</TD>
			<TD>...</TD>
			<TD>...</TD>
		</TR>
	</TABLE>
</CENTER>
<P>
"Start" and "End" are the physical positions (bp) of a gene on the chromosome.
If a gene does not belong to any family, you can specify its family ID to be zero.
</P>
<P>
You can download the above information for whole sequenced genomes from public databases 
of orthologs (e.g. HomoloGene, 
<A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=homologene">
	http://www.ncbi.nlm.nih.gov/sites/entrez?db=homologene
</A>) 
or paralogs (e.g. EPGD, 
<A HREF="http://epgd.biosino.org/EPGD">http://epgd.biosino.org/EPGD</A>).
We have prepared two sample data sets for test here: 
<A HREF="data/orthologs.txt">orthologs</A> and
<A HREF="data/paralogs.txt">paralogs</A>. You can also download the
data from the Ensembl database (see below for details).
</P>

<H3>Download data from Ensembl database</H3>
<P>
Click the button <IMG SRC = "icon/web.png", ALT = "web"> 
and then choose two species you want to compare.
</P>
<P>
We prepared some of the species whose genomes have been completely assembled
in Ensembl database (<A HREF="http://www.ensembl.org/">http://www.ensembl.org</A>). 
CHSMiner will automatically fetch the information through its MySQL client interface.
The speed is determined by your networks. If you want to fetch the data for the
species you are interested in, you can connect the Ensembl public MySQL server 
(see <A HREF="http://www.ensembl.org/info/data/mysql.html">
http://www.ensembl.org/info/data/mysql.html</A>), choose the <I>ensembl_compara_xx </I>
database (where <I>xx</I> is the version), and execute the SQL query
</P>
<P><I>
SELECT stable_id, family_id, display_label, taxon_id, chr_name, chr_strand, chr_start, chr_end<br>
FROM member, family_member<br>
WHERE member.member_id = family_member.member_id AND source_name = 'ENSEMBLGENE' AND taxon_id = ?<br>
</I></P>
<P>
You can find the <I>taxon_id</I> of your species in the NCBI Taxonomy database 
(<A HREF="http://www.ncbi.nlm.nih.gov/taxonomy/">http://www.ncbi.nlm.nih.gov/taxonomy/</A>).
</P>

<P>
Click the button <IMG SRC = "icon/stop.png", ALT = "stop"> to close the 
data set first if you want to change for a new input.
</P>

<A HREF="#INDEX"> <EM>Back to Index</EM> </A>

<CENTER><H2><A NAME="E">Preprocessing</A></H2></CENTER>
<P>
	Click the button <IMG SRC = "icon/properties.png", ALT = "properties"> for preprocessing.
	In the subsequent dialog box, you may have your options for the three tasks below:
</P>
<UL>
	<LI>
	Filter chromosomes you do not want to compare in this analysis. For example, you can 
	remove those incomplete contigs from your data set.
	</LI>
	<LI>
	Decide whether tandem duplicates should be merged. This should usually be yes as tandem
	duplicates will result in redundant information for the homologous regions. <br>
	<B>Note:</B> In this software, we only merge the tandem duplicates without any interrupting genes. 
	The suffix "<I>_cluster</I>" will be added to the symbols of the merged genes. You can also
	define other criteria to merge the tandem duplicates before you input the data.
	</LI>
	<LI>
	Set null family ID, which indicates a gene does not belong to any family. 
	The default value is zero.
	</LI>
</UL>

<A HREF="#INDEX"> <EM>Back to Index</EM> </A>

<CENTER><H2><A NAME="M">CHS detection and evaluation</A></H2></CENTER>
<P>Click the button <IMG SRC = "icon/run.png", ALT = "run"> to start to search for CHSs.</P>
<P>
In the subsequent dialog box, in addition to deciding whether syntenies or paralogons should be identified, 
two critical parameters must be specified:
</P>
<UL>
	<LI>
	The size of a CHS, i.e. the number of distinct families (not genes) 
	linked in the CHS. The default value is two.
	</LI>
	<LI>
	The gap allowed in a CHS, i.e. the number of unmatched genes between two neighboring
	matched ones on the same chromosome. The default value is 30. According to our experience,
	this value is useful in general applications if whole genome data is available. However, you 
	may need to adjust it for special data sets and purposes. 
	</LI>
</UL>
<P>The algorithm is carried out in two stages:</P>
<UL>
	<LI>
	Compare each pair of chromosomes to search for all potential CHSs. A greedy algorithm is 
	used in this stage, i.e. starting from two homologous genes, each at a different 
	chromosomal location, and then looking for two other homologous genes that are each located 
	within the predefined gap from the former two ones. The algorithm has only linear time 
	complexity and is suitable for whole genome analysis.
	</LI>
	<LI>
	Statistical evaluation of the resulted CHSs, i.e. caculating the <I>P</I>-value for a given CHS 
       	under the null hypothesis it is observed from two randomly permutated genomes. We implemented 
	a theoretic algorithm based on the results of Hoberman et al. <A HREF="#2">[2]</A> rather than 
	direct simulation to improve the performance and speed (see <A HREF="#D">more details</A>). 
	</LI>
</UL>
<P>
The process may take a few seconds to minutes determined by your computer and data set.
A progress bar will be shown for users to monitor this process.
</P>
<A HREF="#INDEX"> <EM>Back to Index</EM> </A>

<CENTER><H2><A NAME="P">Output files</A></H2></CENTER>
<P>The output table looks like this:</P>
<CENTER>
	<TABLE ALIGN=CENTER BORDER=1 CELLSPACING=1 CELLPADDING=5>
		<TR>
			<TD><STRONG># Species A</STRONG></TD>
			<TD><STRONG>Chromosome A</STRONG></TD>
			<TD><STRONG>Start A</STRONG></TD>
			<TD><STRONG>End A</STRONG></TD>
			<TD><STRONG>Species B</STRONG></TD>
			<TD><STRONG>Chromosome B</STRONG></TD>
			<TD><STRONG>Start B</STRONG></TD>
			<TD><STRONG>End B</STRONG></TD>
			<TD><STRONG>Size</STRONG></TD>
			<TD><STRONG>P-value</STRONG></TD>
		</TR>
		<TR>
			<TD><STRONG># GeneID A</STRONG></TD>
			<TD><STRONG>Symbol A</STRONG></TD>
			<TD><STRONG>GeneID B</STRONG></TD>
			<TD><STRONG>Symbol B</STRONG></TD>
		</TR>
		<TR>
			<TD>>Mus musculus</TD>
			<TD>14</TD>
			<TD>63616938</TD>
			<TD>65074739</TD>
			<TD>Homo sapiens</TD>
			<TD>8</TD>
			<TD>9949235</TD>
			<TD>12220195</TD>
			<TD>15</TD>
			<TD>7.077E-55</TD>
		</TR>
		<TR>
			<TD> 110265</TD>
			<TD>Msra</TD>
			<TD>4482</TD>
			<TD>MSRA</TD>
		</TR>
		<TR>
			<TD> 71037</TD>
			<TD>4933401F05Rik</TD>
			<TD>203074</TD>
			<TD>UNQ9391</TD>
		</TR>
	</TABLE>
</CENTER>
<P>
The line beginning with ">" is a summary for a CHS:<br>
"A" and "B" are a pair of genomic locations that constitutes the CHS. 
"Size" is the number of distinct families (not genes) linked in the CHS.
"P-value" is the statistical evaluation of the CHS under the null hypothesis
that the gene orders of both genomes are random.
</P>
<P>
Following the line, there is a list containing all the matched genes in the CHS.
</P>
<P>
You can click the button <IMG SRC = "icon/save.png", ALT = "save"> in 'CHS result' pane 
to save the table as text format with items separated by Tab.
</P>
<P>
<B>Note: </B> When several CHSs are tested simultaneously, their raw <I>p</I>-values 
should be corrected. However, we do not implement multiple test correction procedures 
in CHSMiner as there are several methods available (see 
<A HREF = http://privacycore.info/index.php?hl=f5&q=uggc%3A%2F%2Fra.jvxvcrqvn.bet%2Fjvxv%2FZhygvcyr_pbzcnevfbaf>
Multiple comparisons</A>). Although users can choose a proper one in a special case, 
we recommend the strict Bonferroni method, i.e. <BR>
<CENTER><img SRC = "img/corrected_p.gif"></CENTER><BR>
where <I>n</I> is the total number of CHSs tested.
</P>
<A HREF="#INDEX"> <EM>Back to Index</EM> </A>

<CENTER><H2><A NAME="B">CHS visualization</A></H2></CENTER>
<P>CHSMiner can generate two types of images.</P>

<H3>Visualization of whole chromosomal pattern</H3>
<P>
	Click the button <IMG SRC = "icon/link.png", ALT = "link"> in 'CHS result' pane (Fig 3A)
	and select the chromosomes you want to compare in the subsequent dialog box. CHSMiner will render
	all CHSs between a given chromosome and the others (Fig 3B). The image is also interactive:
	you can browse CHSs on the chromosomal view, click on one and jump to the zoomed region (Fig 3C).
</P>

<H3>Visualization of an individual CHS</H3>
<P>
	Select a CHS in the result table (Fig 3A) and then click the button 
	<IMG SRC = "icon/view.png", ALT = "view">. CHSMiner will render and 
	label the homologous genes linked in the CHS (Fig 3C). The image can 
	also be generated from chromosomal view (Fig 3B). For a large CHS, it may
	take a few seconds to render the image.
</P>
<P>
	When the CHS shown is too long to be clear, click the button 
	<IMG SRC = "icon/viewLocal.png", ALT = "viewLocal"> in 'CHS visualization'
	pane to show only part of it (Fig 3D). You can specify the ranges in the 
	subsequent dialog box (measured by gene ranks as you can capture in the image).	
</P>

<CENTER>
	<IMG SRC = "img/CHSVisualization.png", ALT = "Figure 3">
	<P><STRONG>Figure 3 Visualization of CHS</STRONG></P>
</CENTER>

<H3>Graphics operations</H3>
<P>
	Click the button <IMG SRC = "icon/save.png", ALT = "save"> in 'CHS visualization' pane
	to save current image. CHSMiner supports two common types of vector graph formats, i.e. 
	Post Script (.ps) and Scalable Vector Graphics (.svg). These formats are convenient
	for further edit and can easily be converted to bitmap with high resolution.
</P>
<P>
	For user convenience, CHSMiner also provides some other simple graphics operations. Click the 
	buttons <IMG SRC = "icon/back.png", ALT = "back"> and <IMG SRC = "icon/forward.png", ALT = "forward">
	to browse all the images generated. Click the button <IMG SRC = "icon/delete.png", ALT = "delete"> 
	to delete current image. Click the buttons <IMG SRC = "icon/zoomIn.png", ALT = "zoomIn"> and 
	<IMG SRC = "icon/zoomOut.png", ALT = "zoomOut"> to zoom in and out the image. The image can also be 
	dragged by your mouse for a better visualization.
</P>
<A HREF="#INDEX"> <EM>Back to Index</EM> </A>

<CENTER><H2><A NAME="D">More details about statistical evaluation</A></H2></CENTER>
<P>
First, we consider a set of <I>m </I> marked genes (each with different family to exclude tandem repeats) 
forms a cluster with maximum <I>g </I> insertions allowed between any two adjacent genes. 
If we assume every family contains only one gene, the exact probability of observing it in a 
random genome of <I>n</I> genes is <A HREF="#2">[2]</A>:
</P>
<CENTER><img SRC = "img/formular1.gif", ALT = "formular1"></CENTER>
<P>
Next, we consider the general case that a family contains more than one gene. We denote 
<I>F = {f1, f2, ... fm}</I>, where <I>fj</I> is the number of genes of the same family 
with gene <I>j </I> in the cluster. Then the probability above can be corrected as 
<A HREF = "#3">[3]</A>:
</P>
<CENTER><img SRC = "img/formular2.gif", ALT = "formular2"></CENTER>
<P>
The final probability that the cluster can be observed in both genomes of interest, each with 
parameters <I>(n1, F1) </I> and <I>(n2, F2)</I>, can be calculated as
</P>
<CENTER><img SRC = "img/formular3.gif", ALT = "formular3"></CENTER>
<P>, which is the <I>P</I>-value for the given CHS generated from whole genome comparison.</P>
<P>
<B>Note: </B>In the context of whole genome comparison, the difference with Hoberman et al. 
<A HREF="#2">[2]</A> is that we allow the given cluster to be a subset of a larger cluster. 
This processing is more convenient for <I>P</I>-value calculation and the result is more pessimistic.
</P>
<A HREF="#INDEX"> <EM>Back to Index</EM> </A>

<CENTER><H2><A NAME="R">References</A></H2></CENTER>
<OL>
<LI><A NAME="1">
	Simillion, C., K. Vandepoele, and Y. Van de Peer, Recent developments in 
	computational approaches for uncovering genomic homology. Bioessays, 2004. 26(11): p. 1225-35.
	[<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?
		cmd=Retrieve&db=PubMed&dopt=Citation&list_uids=15499578 ">PubMed</A>]
</A></LI>
<LI><A NAME="2">
	Hoberman, R., D. Sankoff, and D. Durand, The statistical analysis of spatially 
	clustered genes under the maximum gap criterion. J Comput Biol, 2005. 12(8): p. 1083-102.
	[<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?
		cmd=Retrieve&db=PubMed&dopt=Citation&list_uids=16241899">PubMed</A>]
</A></LI>
<LI><A NAME="3">
	He, X. and M.H. Goldwasser, Identifying conserved gene clusters in the presence of 
	homology families. J Comput Biol, 2005. 12(6): p. 638-56.
	[<A HREF="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?
		cmd=Retrieve&db=PubMed&dopt=Citation&list_uids=16108708">PubMed</A>]

</A></LI>
</OL>
<A HREF="#INDEX"> <EM>Back to Index</EM></A>


<CENTER><H2><A NAME="L">License statement</A></H2></CENTER>

<H3>Statement of copying permission</H3>
<P>This statement is part of CHSMiner. </P>
<P>
CHSMiner is free software: you can redistribute it and/or modify it under the terms of 
the GNU General Public License (GPL) Version 2 or later as published by the Free Software Foundation.<BR>
CHSMiner is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
See the GNU GPL for more details.<BR>
You should have received a copy of the license along with CHSMiner. If not, 
see <A HREF="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</A>.
</P>

<H3>Statement of third party license</H3>
<P>
For user convenience, <A HREF="http://dev.mysql.com/usingmysql/java/">MySQL Connector/JDBC</A> 
is redistributed along with CHSMiner. The redistribution is licensed under the terms of the GPL.
</P>
<P>
For user convenience, <A HREF="http://xmlgraphics.apache.org/batik">Batik SVG Toolkit</A> 
is redistributed along with CHSMiner. The redistribution is licensed under the terms of 
the Apache License, Version 2.0. You may not use CHSMiner except in compliance with the License. 
You should have received a copy of the license along with CHSMiner. If not, see 
<A HREF="http://www.apache.org/licenses/LICENSE-2.0"> http://www.apache.org/licenses/LICENSE-2.0</A>.
</P>
<A HREF="#INDEX"> <EM>Back to Index</EM></A>

<CENTER><H2><A NAME="C">Contact us</A></H2></CENTER>
<H4>Zhen Wang</H4>
<P>
Bioinformatics Center, Key Lab of Systems Biology, Shanghai Institutes for Biological Sciences, 
Chinese Academy of Sciences, P.R.China <BR>
Email: <A href="mailto:zhenwang100@gmail.com">zhenwang100@gmail.com</A>
</P>
<H4>Yixue Li</H4>
<P>
Bioinformatics Center, Key Lab of Systems Biology, Shanghai Institutes for Biological Sciences, 
Chinese Academy of Sciences, P.R.China <Br>
Email: <a href="mailto:yxli@sibs.ac.cn">yxli@sibs.ac.cn</a>
</p>
<A HREF="#INDEX"> <EM>Back to Index</EM></A>

<HR>
<CENTER>	
<A href="http://www.sibs.ac.cn/en/"><img src="img/sibs.jpg" alt="SIBS" width="150" height="50"/></A>
	&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<A href="http://www.sysbio.ac.cn/"><img src="img/sysbio.jpg" alt="SYSBIO" width="50" height="50"/></A>
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<A href="http://www.biosino.org/"><img src="img/biosino.gif" alt="BIOSINO" width="150" height="50"/></a>
</CENTER>
<SMALL><P>
	Copyright (C) 2008 Bioinformatics Center, Shanghai Institutes for Biological Sciences, 
	Chinese Academy of Sciences. All rights reserved. 
</P></SMALL>

